rfc/src/rfcs/0004-loop-contract-synthesis.md

```diff
@@ -0,0 +1,209 @@
+- **Feature Name:** Loop-contract synthesis
+- **Feature Request Issue:** <https://github.com/model-checking/kani/issues/2214>
+- **RFC PR:** <https://github.com/model-checking/kani/pull/2215>
+- **Status:** Under Review
+- **Version:** 0
+- **Proof-of-concept:** <https://github.com/qinheping/kani/tree/kani-synthesizer>
+
+## Summary
+
+A new option that allows users to verify programs without unwinding loops by synthesizing loop contracts for those loops.
+
+
+## User Impact
+Currently Kani does not support proving correctness (i.e. assertions never fail) on programs with unbounded control flow (e.g. loops with dynamic bounds). Kani unrolls all unbounded loops until a global threshold (unwinding number) specified by the user and then verifies this unrolled program, which limits the set of programs it can verify.
+
+A new Kani flag `--synthesize-loop-contracts` will be created that can be used to enable the goto-level loop-contract synthesizer `goto-synthesizer`.
+The idea of [loop contracts](https://arxiv.org/pdf/2010.05812.pdf) is, instead of unwinding loops, we abstracts those loops as non-loop structures that can cover arbitrary iterations of the loops.
+The loop contract synthesizer, when enabled, will attempt to synthesize loop contracts for all loops.
+CBMC can then apply the synthesized loop contracts and verify the program without unwinding any loop.
+So, the synthesizer will help to verify the programs that require Kani to unwind loops for a very large number of times to cover all iterations.
+
+
+For example, the number of executed iterations of the loop in the following harness is dynamically bounded by an unbounded variable `y`.
+Only an unwinding value of `i32::MAX` can guarantee to cover all iteration of the loop, and hence satisfies the unwinding assertions.
```

Comment:
```suggestion
Only an unwinding value of `i32::MAX` can guarantee to cover all iterations of the loop, and hence satisfy the unwinding assertions.
```
? Not sure about "satisfies -> satisfy", but I think you're talking about the unwinding value.